Avalonia.Labs.Gif/GifImage.cs

```diff
@@ -0,0 +1,209 @@
+        Update();
+        base.OnAttachedToVisualTree(e);
+    }
+
+    protected override void OnDetachedFromVisualTree(VisualTreeAttachmentEventArgs e)
+    {
+        var compositor = ElementComposition.GetElementVisual(this)?.Compositor;
+        if (compositor == null || _customVisual?.Compositor == compositor)
+            return;
+            
+        ElementComposition.SetElementChildVisual(this, null);
+        _customVisual?.SendHandlerMessage(GifCustomVisualHandler.StopMessage);
+        _customVisual = null;
+    }
+
+
+    private void Update()
+    {
+        if (_customVisual is null || _gifInstance is null)
+            return;
+
+        var dpi = this.GetVisualRoot()?.RenderScaling ?? 1.0;
+        var sourceSize = _gifInstance.GifPixelSize.ToSize(dpi);
+        var viewPort = new Rect(Bounds.Size);
+
+        var scale = Stretch.CalculateScaling(Bounds.Size, sourceSize, StretchDirection);
+        var scaledSize = sourceSize * scale;
+        var destRect = viewPort
+            .CenterRect(new Rect(scaledSize))
+            .Intersect(viewPort);
+
+        _customVisual.Size = Stretch == Stretch.None ?
+            new Vector2((float)sourceSize.Width, (float)sourceSize.Height) : 
+            new Vector2((float)destRect.Size.Width, (float)destRect.Size.Height);
+
+        _customVisual.Offset = new Vector3((float)destRect.Position.X, (float)destRect.Position.Y, 0);
+    }
+
+    /// <summary>
+    /// Measures the control.
+    /// </summary>
+    /// <param name="availableSize">The available size.</param>
+    /// <returns>The desired size of the control.</returns>
+    protected override Size MeasureOverride(Size availableSize)
+    {
+        var result = new Size();
+        var scaling = this.GetVisualRoot()?.RenderScaling ?? 1.0;
+        if (_gifInstance != null)
+        {
+            result = Stretch.CalculateSize(availableSize, _gifInstance.GifPixelSize.ToSize(scaling),
+                StretchDirection);
+        }
+
+        return result;
+    }
+
+    /// <inheritdoc/>
+    protected override Size ArrangeOverride(Size finalSize)
+    {
+        if (_gifInstance is null) return new Size();
+        var scaling = this.GetVisualRoot()?.RenderScaling ?? 1.0;
+        var sourceSize = _gifInstance.GifPixelSize.ToSize(scaling);
+        var result = Stretch.CalculateSize(finalSize, sourceSize);
+        return result;
+    }
+
+
+    private void SourceChanged(AvaloniaPropertyChangedEventArgs e)
+    {
+        if (e.NewValue is null ||
+            (e.NewValue is string value && !Uri.IsWellFormedUriString(value, UriKind.Absolute)))
+        {
+            return;
+        }
+
+        if (_customVisual is null)
+        {
+            _initialSource = e.NewValue;
+            return;
+        }
+
+        UpdateGifInstance(e.NewValue);
+
+        InvalidateArrange();
+        InvalidateMeasure();
+        Update();
+    }
+
+    private void UpdateGifInstance(object source)
+    {
+        _gifInstance?.Dispose();
```

Comment:
That might be a reason of exception you mentioned in the chat.

GifInstance is currently accessed from both UI and Render threads. 
Where it can be a situation when git instance is currently used on Render thread, but suddenly disposed on UI thread.

Ideally I would recommend moving 100% of GitInstance to the Render thread visual handler. And only update it via Messages API.

Reply:
Alternatively, you can add a bunch of `locks` all over gif instance.